### PR TITLE
Set leader-elect for kube-scheduler to true / Adds new configuration types and rework how we set defaults

### LINKED ISSUE #59729: kube-scheduler defaults --lead-elect to false whereas kube-controller-manager defaults to true
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
[Building High-Availability Clusters](https://kubernetes.io/docs/admin/high-availability/building/) docs recommend using '--leader-elect' to have HA setup for kube-scheduler and kube-controller-manager. However, these two components approach differently what is default value for the flag. [kube-controller-manager sets it to true](https://kubernetes.io/docs/reference/generated/kube-controller-manager/). [kube-schedulers sets it to false](https://kubernetes.io/docs/reference/generated/kube-scheduler/).

Seems that setting the value to true in kube-scheduler code was lost in refactoring scheduler API config in efb2bb71cd. The code before refactoring had the value [set to true](https://github.com/kubernetes/kubernetes/commit/efb2bb71cd4dede007289634bfe3c51484625ba9#diff-809dde23edc35bf4c46cdfea08e9abdcL62).

**What you expected to happen**:

kube-scheduler set true as a default value for '--leader-elect'

**Environment**:
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1", GitCommit:"3a1c9449a956b6026f075fa3134ff92f7d55f812", GitTreeState:"archive", BuildDate:"2018-01-10T13:28:21Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1", GitCommit:"3a1c9449a956b6026f075fa3134ff92f7d55f812", GitTreeState:"archive", BuildDate:"2018-01-10T13:28:21Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

```

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:
Thanks to some great sleuthing by ikruglov!

kube-controller-manager defaults --leader-elect to true. We should
do the same for kube-scheduler. kube-scheduler used to have this
set to true, but it got lost during refactoring in:
efb2bb71cd4dede007289634bfe3c51484625ba9

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #59729

**Special notes for your reviewer**:

**Release note**:

```release-note
kube-scheduler has been fixed to use `--leader-elect` option back to true (as it was in previous versions)
```

### CODE REVIEW

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

File: cmd/kube-scheduler/app/server.go
Comment (liggitt) [inline cmd/kube-scheduler/app/server.go]: seems like defaults should be applied in defaults.go, not here (and options round-tripped through those defaults)

**Comment by liggitt** [comment]
why is this default different than all the other defaults for the scheduler config, applied in the api defaulting path? the setting here won't get applied when reading from file

File: cmd/kube-scheduler/app/server.go
Comment (dims) [reply cmd/kube-scheduler/app/server.go]: i was trying to match what is done in kube-controller ... will fix https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/options/options.go#L52

**Comment by dims** [comment]
@liggitt looks like we need to do both ... cmd/kube-scheduler/app/server.go and pkg/apis/componentconfig/v1alpha1/defaults.go if i don't patch up server.go, then the command line default is still false.

**Comment by liggitt** [comment]
> if i don't patch up server.go, then the command line default is still false.

that means the internal config is not being round tripped and defaulted... that seems like it should be fixed as well

**Comment by dims** [comment]
@liggitt can you point me to an existing pattern/files where this is done please?

**Comment by liggitt** [comment]
> @liggitt can you point me to an existing pattern/files where this is done please?

for kubeproxy:

https://github.com/kubernetes/kubernetes/blob/fdeaa8c67a5a5f63626f6b0399b158dd0ce36dcd/cmd/kube-proxy/app/server.go#L303-L343

for kubelet:

https://github.com/kubernetes/kubernetes/blob/fdeaa8c67a5a5f63626f6b0399b158dd0ce36dcd/cmd/kubelet/app/options/options.go#L255-L268

**Comment by dims** [comment]
Thanks @liggitt looks like it was already setup. but there was a mismatch in `bool` vs `*bool` for LeaderElect (in the 2 types.go file) and a hard coding of LeaderElect to false in DefaultLeaderElectionConfiguration.

File: pkg/apis/componentconfig/types.go
Comment (liggitt) [inline pkg/apis/componentconfig/types.go]: internal `bool`, external `*bool` was intentional. Convert_Pointer_bool_To_bool handles the conversion, and external pointer lets you apply defaulting logic if nil.

Comment (dims) [reply pkg/apis/componentconfig/types.go]: Ack going back to what i had before

**Comment by liggitt** [comment]
/hold cancel
LGTM

**Comment by bsalamat** [comment]
The changes LGTM as well, but I let @timothysc take a look as he is more familiar with the learder election part.

**Comment by dims** [comment]
@timothysc this is ready now. please take a look. @liggitt helped me whip it into shape

**Comment by dims** [comment]
@mtaufen PTAL, needs your approval for pkg/apis/componentconfig/OWNERS

File: pkg/client/leaderelectionconfig/config.go
Comment (mtaufen) [inline pkg/client/leaderelectionconfig/config.go]: Shouldn't the defaults be produced by running the defaulter on a raw object? This function can easily skew from what the defaulter produces, and the defaulter should be the source of truth.

Additionally, there should not be a concept of an internal object having defaults. Defaults are part of versioned APIs.

I'd expect something like:
```
versioned := &componentconfigv1alpha1.LeaderElectionConfiguration{}
scheme.Default(versioned)
internal := &componentconfig.LeaderElectionConfiguration{}
if err := scheme.Convert(versioned, internal, nil); err != nil {
	return nil, err
}
return internal, nil
```

Comment (dims) [reply pkg/client/leaderelectionconfig/config.go]: @mtaufen Done!

**Comment by dims** [comment]
@thockin @luxas @wlan0 - Can one of you please approve for cmd/cloud-controller-manager?
@mikedanese @vishh @mtaufen - Can one of you please approve for pkg/apis/componentconfig?

File: pkg/client/leaderelectionconfig/config.go
Comment (mtaufen) [reply pkg/client/leaderelectionconfig/config.go]: You don't have to do it in this PR, but I noticed the pattern of "defaults" divorced from the defaulters is also shared by `k8s.io/kubernetes/cmd/controller-manager/app/options.NewDefaultControllerManagerComponentConfig`, so it would be a good thing to clean that one up too.

Comment (dims) [reply pkg/client/leaderelectionconfig/config.go]: Ack @mtaufen i got the hang of it, will give it a shot in another PR

Comment (mtaufen) [reply pkg/client/leaderelectionconfig/config.go]: Hmm, actually, now that I look more closely, I guess `LeaderElectionConfiguration` is just a substructure of the `KubeControllerManagerConfiguration`?

It is a bit strange to have a defaulter for a substructure like this. Without the context of *which* component it belongs to, it's premature to set a default. Do we need to merge this PR asap to fix failing tests or can we clean up the defaulting for `KubeControllerManagerConfiguration` here too?

Comment (dims) [reply pkg/client/leaderelectionconfig/config.go]: @mtaufen Done! i've cleaned up the defaulting of `KubeControllerManagerConfiguration`

**Comment by dims** [comment]
@mtaufen @timothysc can you please add this to the 1.10 milestone?

File: cmd/controller-manager/app/options/options.go
Comment (mtaufen) [inline cmd/controller-manager/app/options/options.go]: Did you make sure that all these defaults match the ones applied by the defaulter?

File: pkg/apis/componentconfig/v1alpha1/conversion.go
Comment (mtaufen) [inline pkg/apis/componentconfig/v1alpha1/conversion.go]: what prevented us from generating these conversions?

Comment (dims) [reply pkg/apis/componentconfig/v1alpha1/conversion.go]: @mtaufen i saw some warnings in zz_generated.conversion.go. Are those the things that prevented us from fully auto generating? 

```
zz_generated.conversion.go:	// WARNING: in.LeaderElection requires manual conversion: inconvertible types (*k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.LeaderElectionConfiguration vs k8s.io/kubernetes/pkg/apis/componentconfig.LeaderElectionConfiguration)
zz_generated.conversion.go:	// WARNING: in.VolumeConfiguration requires manual conversion: inconvertible types (*k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.VolumeConfiguration vs k8s.io/kubernetes/pkg/apis/componentconfig.VolumeConfiguration)
zz_generated.conversion.go:	// WARNING: in.LeaderElection requires manual conversion: inconvertible types (k8s.io/kubernetes/pkg/apis/componentconfig.LeaderElectionConfiguration vs *k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.LeaderElectionConfiguration)
zz_generated.conversion.go:	// WARNING: in.VolumeConfiguration requires manual conversion: inconvertible types (k8s.io/kubernetes/pkg/apis/componentconfig.VolumeConfiguration vs *k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.VolumeConfiguration)
zz_generated.conversion.go:	// WARNING: in.PersistentVolumeRecyclerConfiguration requires manual conversion: inconvertible types (*k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.PersistentVolumeRecyclerConfiguration vs k8s.io/kubernetes/pkg/apis/componentconfig.PersistentVolumeRecyclerConfiguration)
zz_generated.conversion.go:	// WARNING: in.PersistentVolumeRecyclerConfiguration requires manual conversion: inconvertible types (k8s.io/kubernetes/pkg/apis/componentconfig.PersistentVolumeRecyclerConfiguration vs *k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.PersistentVolumeRecyclerConfiguration)
```

Comment (mtaufen) [reply pkg/apis/componentconfig/v1alpha1/conversion.go]: Do the substructures need to be pointers in the v1alpha1 type? If not I think you can avoid manual conversions.

Comment (dims) [reply pkg/apis/componentconfig/v1alpha1/conversion.go]: Ack. Let me try that @mtaufen

File: cmd/controller-manager/app/options/options.go
Comment (mtaufen) [inline cmd/controller-manager/app/options/options.go]: Since componentconfig is installed (`k8s.io/kubernetes/pkg/apis/componentconfig/install/install.go`) in legacyscheme (`k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go`), you can probably just use legacyscheme.Scheme instead of creating a new one here.

File: cmd/controller-manager/app/options/options.go
Comment (mtaufen) [inline cmd/controller-manager/app/options/options.go]: this feels out of place here, why not in the defaulter?

File: pkg/apis/componentconfig/v1alpha1/defaults_test.go
Comment (mtaufen) [inline pkg/apis/componentconfig/v1alpha1/defaults_test.go]: maybe `TestControllerDefaultsRoundTrip`?

File: pkg/client/leaderelectionconfig/config.go
Comment (mtaufen) [inline pkg/client/leaderelectionconfig/config.go]: same comment here about probably not needing to create a new scheme, because componentconfig should be installed in legacyscheme

File: cmd/controller-manager/app/options/options.go
Comment (dims) [reply cmd/controller-manager/app/options/options.go]: good one! will try that.

File: cmd/controller-manager/app/options/options.go
Comment (dims) [reply cmd/controller-manager/app/options/options.go]: this is called from 2 places one from cloud-controller-manager and another from kube-controller-manager with different values for insecurePort. so the default depends on the binary essentially

File: pkg/apis/componentconfig/v1alpha1/defaults_test.go
Comment (dims) [reply pkg/apis/componentconfig/v1alpha1/defaults_test.go]: fixed

File: pkg/client/leaderelectionconfig/config.go
Comment (dims) [reply pkg/client/leaderelectionconfig/config.go]: actually removed this method since we no longer use it in options.go.

File: cmd/controller-manager/app/options/options.go
Comment (dims) [reply cmd/controller-manager/app/options/options.go]: no dice with `legacyscheme.Scheme`, switched back to creating a new scheme

**Comment by liggitt** [comment]
while I'm generally in favor of these changes... this is a large restructure coming in very late. can we do the minimal "default true" change for code freeze and follow up with this PR?

File: cmd/controller-manager/app/options/options.go
Comment (mtaufen) [reply cmd/controller-manager/app/options/options.go]: It might be better not to parameterize it here, and to instead independently override it in code specific to the independent controller-managers.

File: cmd/controller-manager/app/options/options.go
Comment (mtaufen) [reply cmd/controller-manager/app/options/options.go]: hmm, what failed?

**Comment by dims** [comment]
@liggitt ack, going back to my first attempt in this PR :) please see https://github.com/kubernetes/kubernetes/pull/60524

**Comment by dims** [comment]
we merged a small change and pushing this PR off to v1.11

**Comment by dims** [comment]
@mtaufen @liggitt let's get this in please? now that master is open for 1.11. thanks!

**Comment by dims** [comment]
@liggitt now that master is open and it's early in the cycle, let's please get this in.

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/types.go]: Why does this need type info? It's not a top level type, right?

File: cmd/controller-manager/app/options/options.go
Comment (liggitt) [reply cmd/controller-manager/app/options/options.go]: new scheme is preferred anyway, rather than polluting a shared s heme with config specific types

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (dims) [reply pkg/apis/componentconfig/v1alpha1/types.go]: @liggitt some of the tests fail without it

File: cmd/controller-manager/app/options/options.go
Comment (liggitt) [reply cmd/controller-manager/app/options/options.go]: that is probably a sign the config for those two components isn't correct yet. Most of the values don't apply to the cloud controller manager, so I'd probably expect it to get its own config object

File: cmd/controller-manager/app/options/options.go
Comment (liggitt) [reply cmd/controller-manager/app/options/options.go]: Something we've found helpful with similar setups in openshift is having a sanity check unitntest that constructs a config struct, populates it from defaults (or calls this method), then DeepEqual compares it to an expected struct. Setting that unit test up against master, then making sure it passes with no changes against this PR would give me confidence we're not changing anything here. 

That also catches drift or accidental removal of defaults, which is important with config types that are in the process of being moved and are currently shared.

File: cmd/controller-manager/app/options/options.go
Comment (dims) [reply cmd/controller-manager/app/options/options.go]: i can tweak this in a subsequent PR

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (liggitt) [reply pkg/apis/componentconfig/v1alpha1/types.go]: unless this is serialized in standalone way (which I'm not aware of), this should not have type info. tests that fail without that should be fixed.

File: pkg/apis/componentconfig/v1alpha1/defaults.go
Comment (liggitt) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: still outstanding

Comment (dims) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: ack deleting now

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (dims) [reply pkg/apis/componentconfig/v1alpha1/types.go]: ack removing it

File: pkg/apis/componentconfig/types.go
Comment (dims) [reply pkg/apis/componentconfig/types.go]: ack removing it

File: cmd/controller-manager/app/options/options.go
Comment (dims) [reply cmd/controller-manager/app/options/options.go]: @liggitt there's a `TestAddFlags` in options_test.go in this directory. it has not been touched. it does use `NewCloudControllerManagerOptions`

**Comment by dims** [comment]
Ready and all green now @liggitt . thanks!

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/types.go]: these should probably end up in `controllermanager.config.k8s.io` under `pkg/controller/apis`, similar to kubelet. can you move them in a follow up?

File: cmd/controller-manager/app/options/options.go
Comment (liggitt) [reply cmd/controller-manager/app/options/options.go]: > there's a TestAddFlags in options_test.go in this directory. it has not been touched. it does use NewCloudControllerManagerOptions

that's good, but that test is making sure flags override defaults, not ensuring the defaults match what is expected. if that test also compared against expected defaults before binding/parsing flags, that would cover it

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (dims) [reply pkg/apis/componentconfig/v1alpha1/types.go]: yes i can

File: cmd/controller-manager/app/options/options.go
Comment (dims) [reply cmd/controller-manager/app/options/options.go]: cool. will do that shortly

Comment (dims) [reply cmd/controller-manager/app/options/options.go]: Done. added a new test without parsing/binding flags. also ran it against master - http://paste.openstack.org/show/718026/ - no errors.

**Comment by liggitt** [comment]
thanks
/lgtm

**Comment by bsalamat** [comment]
Could you please update the title of this PR and state that it adds new configuration types?

**Pull request merged**
